src/Silex/EventListener/LocaleListener.php

```diff
@@ -24,24 +28,106 @@
  *
  * @author Fabien Potencier <fabien@symfony.com>
  */
-class LocaleListener extends BaseLocaleListener
+class LocaleListener implements EventSubscriberInterface
 {
+    /**
+     * @var \Silex\Application
+     */
     protected $app;
 
+    /**
+     * @var \Symfony\Component\Routing\RequestContextAwareInterface|null
+     */
+    private $router;
+
+    /**
+     * @var string
+     */
+    private $defaultLocale;
+
+    /**
+     * @var \Symfony\Component\HttpFoundation\RequestStack|null
+     */
+    private $requestStack;
+
     public function __construct(Application $app, ?RequestContextAwareInterface $router = null, ?RequestStack $requestStack = null)
     {
-        parent::__construct($requestStack, $app['locale'], $router);
-
         $this->app = $app;
+
+        $this->defaultLocale = $app['locale'];
+        $this->requestStack = $requestStack;
+        $this->router = $router;
     }
 
     /**
      * @return void
      */
-    public function onKernelRequest(GetResponseEvent $event)
+    public function setDefaultLocale(KernelEvent $event)
     {
-        parent::onKernelRequest($event);
+        $event->getRequest()->setDefaultLocale($this->defaultLocale);
+    }
+
+    /**
+     * @return void
+     */
+    public function onKernelRequest(RequestEvent $event)
+    {
+        $request = $event->getRequest();
+
+        $this->setLocale($request);
+        $this->setRouterContext($request);
 
         $this->app['locale'] = $event->getRequest()->getLocale();
     }
+
+    /**
+     * @param \Symfony\Component\HttpKernel\Event\FinishRequestEvent $event
+     *
+     * @return void
+     */
+    public function onKernelFinishRequest(FinishRequestEvent $event)
+    {
+        if (null !== $parentRequest = $this->requestStack->getParentRequest()) {
+            $this->setRouterContext($parentRequest);
+        }
+    }
+
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return void
+     */
+    private function setLocale(Request $request)
+    {
+        if ($locale = $request->attributes->get('_locale')) {
+            $request->setLocale($locale);
+        }
+    }
+
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return void
+     */
+    private function setRouterContext(Request $request)
+    {
+        if ($this->router !== null) {
+            $this->router->getContext()->setParameter('_locale', $request->getLocale());
+        }
+    }
+
+    /**
+     * @return array[][]
```

Comment:
Just `array` might suffice here, as this one here doesnt provide much additional value.